### ISSUE TITLE
Unable to export to PDF--"Error: Functions that produce HTML output found in document targeting latex output"

### ISSUE THREAD

**Issue by isaactpetersen** [rstudio/bookdown#1119]
I'm experiencing an issue similar to the issue here (it is closed even though some people's issues did not appear to be resolved): https://github.com/rstudio/bookdown/issues/440

Basically, I'm trying to output my book (**_not_** the `bookdown-demo` or `bookdown` examples) to PDF using `bookdown`, but receive the following error:

```
Error: Functions that produce HTML output found in document targeting latex output.
Please change the output type of this document to HTML. Alternatively, you can allow
HTML output in non-HTML formats by adding this option to the YAML front-matter of
your rmarkdown file:

  always_allow_html: true

Note however that the HTML output will not be visible in non-HTML formats.
```

I tried what was suggested [here](https://github.com/rstudio/bookdown/issues/440), and ran the following:

```
install.packages("webshot")
webshot::install_phantomjs(force = TRUE)
```

Note that I am successfully able to output my book to HTML, but not to PDF. Also note that I am successfully able to output the `bookdown-demo` (https://github.com/rstudio/bookdown-demo) to PDF, so it's something specific to the content of my book. The other issue said not to add `always_allow_html: true` so I didn't do that, but when I tried that in the past, I got errors similar to this one: https://github.com/rstudio/bookdown/issues/440#issuecomment-419117625

Here's some information for reproducibility:

Here are the contents of the top of `index.rmd`:
```
--- 
title: "My Title"
author: "My Name"
date: "`r Sys.Date()`"
site: bookdown::bookdown_site
documentclass: book
bibliography: [book.bib, packages.bib]
csl: apa.csl
link-citations: yes
description: "My description."
github-repo:
cover-image: "Images/cover.png"
---
```

Here are the contents of `_bookdown.yml`:
```
book_filename: "Book"
delete_merged_file: true
language:
  ui:
    chapter_name: "Chapter "
```

Here are the contents of `_output.yml`:
```
bookdown::gitbook:
  split_by: chapter #split_by: none #change to 'none' to save entire book to a single html file
  css: style.css
  toc_depth: 5
  config:
    toc:
      collapse: section
      before: |
        <li><a href="./">My Title</a></li>
      after: |
        <li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>
    download: ["pdf", "epub"]
bookdown::pdf_book:
  includes:
    in_header: preamble.tex
  latex_engine: xelatex
  citation_package: natbib
  keep_tex: yes
bookdown::epub_book: default
```

Here's my session info:
``` r
packageVersion("webshot")
#> [1] '0.5.2'

system('phantomjs --version')
#> 2.1.1
#> [1] 0

rmarkdown::pandoc_version()
#> [1] '2.11.2'

system('pdflatex --version')
#> pdflatex: security risk: running with elevated privileges
#> MiKTeX-pdfTeX 4.0.1 (MiKTeX 20.6.29)
#> Â© 1982 D. E. Knuth, Â© 1996-2020 HÃ n Tháº¿ ThÃ nh
#> TeX is a trademark of the American Mathematical Society.
#> using bzip2 version 1.0.6, 6-Sept-2010
#> compiled with curl version 7.61.1; using libcurl/7.61.1 WinSSL
#> compiled with expat version 2.2.6; using expat_2.2.6
#> compiled with jpeg version 9.3
#> compiled with liblzma version 50020042; using 50020042
#> compiled with libpng version 1.6.37; using 1.6.37
#> compiled with libressl version LibreSSL 2.8.2; using LibreSSL 2.8.2
#> compiled with MiKTeX Application Framework version 4.0; using 4.0
#> compiled with MiKTeX Core version 4.0; using 4.0
#> compiled with MiKTeX Archive Extractor version 4.0; using 4.0
#> compiled with MiKTeX Package Manager version 4.0; using 4.0
#> compiled with poppler version 0.60.1
#> compiled with uriparser version 0.9.2
#> compiled with zlib version 1.2.11; using 1.2.11
#> [1] 0

sessionInfo()
#> R version 4.0.2 (2020-06-22)
#> Platform: x86_64-w64-mingw32/x64 (64-bit)
#> Running under: Windows 10 x64 (build 19041)
#> 
#> Matrix products: default
#> 
#> Random number generation:
#>  RNG:     L'Ecuyer-CMRG 
#>  Normal:  Inversion 
#>  Sample:  Rejection 
#>  
#> locale:
#> [1] LC_COLLATE=English_United States.1252  LC_CTYPE=English_United States.1252    LC_MONETARY=English_United States.1252
#> [4] LC_NUMERIC=C                           LC_TIME=English_United States.1252    
#> 
#> attached base packages:
#>  [1] grid      stats4    splines   stats     graphics  grDevices utils     datasets  methods   base     
#> 
#> other attached packages:
#>  [1] reprex_1.0.0            mirtCAT_1.10            shiny_1.6.0             dagitty_0.3-1           nFactors_2.4.1         
#>  [6] corrplot_0.84           dmacs_0.1.0             msir_1.3.3              ggpubr_0.4.0            gridExtra_2.3          
#> [11] uroc_0.1.0              PredictABEL_1.2-4       ResourceSelection_0.3-5 rms_6.2-0               SparseM_1.81           
#> [16] Hmisc_4.5-0             Formula_1.2-4           survival_3.2-10         pROC_1.17.0.1           magrittr_2.0.1         
#> [21] mirt_1.33.2             lattice_0.20-41         drc_3.0-1               semTools_0.5-4          quantreg_5.85          
#> [26] mice_3.13.0             snow_0.4-3              simsem_0.5-16           DT_0.17                 MASS_7.3-53.1          
#> [31] kableExtra_1.3.4        rockchalk_1.8.144       semPlot_1.1.2           MOTE_1.0.2              tinytex_0.30           
#> [36] gtheory_0.1.2           lme4_1.1-26             Matrix_1.3-2            psychmeta_2.4.2         lavaan_0.6-8           
#> [41] MBESS_4.8.0             blandr_0.5.1            psych_2.1.3             viridis_0.5.1           viridisLite_0.3.0      
#> [46] forcats_0.5.1           stringr_1.4.0           dplyr_1.0.5             purrr_0.3.4             readr_1.4.0            
#> [51] tidyr_1.1.3             tibble_3.1.0            ggplot2_3.3.3           tidyverse_1.3.0         bookdown_0.21          
#> [56] knitr_1.31              rmarkdown_2.7          
#> 
#> loaded via a namespace (and not attached):
#>   [1] fortunes_1.5-4        estimability_1.3      coda_0.19-4           multcomp_1.4-16       styler_1.4.0          data.table_1.14.0    
#>   [7] rpart_4.1-15          generics_0.1.0        callr_3.6.0           TH.data_1.0-10        polspline_1.1.19      webshot_0.5.2        
#>  [13] xml2_1.3.2            lubridate_1.7.10      httpuv_1.5.5          assertthat_0.2.1      xfun_0.22             jquerylib_0.1.3      
#>  [19] hms_1.0.0             evaluate_0.14         promises_1.2.0.1      fansi_0.4.2           progress_1.2.2        dbplyr_2.1.0         
#>  [25] readxl_1.3.1          igraph_1.2.6          DBI_1.1.1             tmvnsim_1.0-2         Rsolnp_1.16           htmlwidgets_1.5.3    
#>  [31] reshape_0.8.8         ellipsis_0.3.1        backports_1.2.1       V8_3.4.0              pbivnorm_0.6.0        permute_0.9-5        
#>  [37] jmvcore_1.2.23        vctrs_0.3.7           ROCR_1.0-11           abind_1.4-5           withr_2.4.1           PBSmodelling_2.68.8  
#>  [43] checkmate_2.0.0       emmeans_1.5.5-1       vegan_2.5-7           fdrtool_1.2.16        prettyunits_1.1.1     mclust_5.4.7         
#>  [49] mnormt_2.0.2          svglite_2.0.0         cluster_2.1.1         mi_1.0                crayon_1.4.1          pkgconfig_2.0.3      
#>  [55] labeling_0.4.2        nlme_3.1-152          nnet_7.3-15           rlang_0.4.10          lifecycle_1.0.0       MatrixModels_0.5-0   
#>  [61] sandwich_3.0-0        kutils_1.70           modelr_0.1.8          cellranger_1.1.0      tcltk_4.0.2           matrixStats_0.58.0   
#>  [67] regsem_1.6.2          carData_3.0-4         boot_1.3-27           zoo_1.8-9             base64enc_0.1-3       processx_3.5.0       
#>  [73] png_0.1-7             arm_1.11-2            jpeg_0.1-8.1          rstatix_0.7.0         ggsignif_0.6.1        scales_1.1.1         
#>  [79] lpSolve_5.6.15        plyr_1.8.6            compiler_4.0.2        RColorBrewer_1.1-2    plotrix_3.8-1         cli_2.3.1            
#>  [85] ps_1.6.0              pbapply_1.4-3         htmlTable_2.1.0       mgcv_1.8-34           tidyselect_1.1.0      stringi_1.5.3        
#>  [91] lisrelToR_0.1.4       sem_3.1-11            highr_0.8             yaml_2.2.1            OpenMx_2.19.1         latticeExtra_0.6-29  
#>  [97] sass_0.3.1            polynom_1.4-0         tools_4.0.2           animation_2.6         parallel_4.0.2        rio_0.5.26           
#> [103] matrixcalc_1.0-3      rstudioapi_0.13       foreign_0.8-81        GPArotation_2014.11-1 farver_2.1.0          digest_0.6.27        
#> [109] Rcpp_1.0.6            car_3.0-10            broom_0.7.5           metafor_2.4-0         ez_4.4-0              later_1.1.0.1        
#> [115] httr_1.4.2            Deriv_4.1.3           colorspace_2.0-0      rvest_1.0.0           XML_3.99-0.6          fs_1.5.0             
#> [121] truncnorm_1.0-8       statmod_1.4.35        conquer_1.0.2         systemfonts_1.0.1     xtable_1.8-4          jsonlite_1.7.2       
#> [127] nloptr_1.2.2.2        corpcor_1.6.9         glasso_1.11           dcurver_0.9.2         R6_2.5.0              pillar_1.5.1         
#> [133] htmltools_0.5.1.1     mime_0.10             glue_1.4.2            fastmap_1.1.0         minqa_1.2.4           codetools_0.2-18     
#> [139] mvtnorm_1.1-1         utf8_1.2.1            bslib_0.2.4           curl_4.3              gtools_3.8.2          clipr_0.7.1          
#> [145] zip_2.1.1             openxlsx_4.2.3        qgraph_1.6.9          munsell_0.5.0         haven_2.3.1           reshape2_1.4.4       
#> [151] gtable_0.3.0         

xfun::session_info('bookdown')
#> R version 4.0.2 (2020-06-22)
#> Platform: x86_64-w64-mingw32/x64 (64-bit)
#> Running under: Windows 10 x64 (build 19041), RStudio 1.4.1103
#> 
#> Random number generation:
#>  RNG:     L'Ecuyer-CMRG 
#>  Normal:  Inversion 
#>  Sample:  Rejection 
#>  
#> Locale:
#>   LC_COLLATE=English_United States.1252  LC_CTYPE=English_United States.1252    LC_MONETARY=English_United States.1252
#>   LC_NUMERIC=C                           LC_TIME=English_United States.1252    
#> 
#> Package version:
#>   base64enc_0.1.3   bookdown_0.21     digest_0.6.27     evaluate_0.14     glue_1.4.2        graphics_4.0.2    grDevices_4.0.2  
#>   highr_0.8         htmltools_0.5.1.1 jsonlite_1.7.2    knitr_1.31        magrittr_2.0.1    markdown_1.1      methods_4.0.2    
#>   mime_0.10         rlang_0.4.10      rmarkdown_2.7     stats_4.0.2       stringi_1.5.3     stringr_1.4.0     tinytex_0.30     
#>   tools_4.0.2       utils_4.0.2       xfun_0.22         yaml_2.2.1       
```

Thanks in advance!

## Checklist

When filing a _bug report_, please check the boxes below to confirm that you have provided us with the information we need. Have you:

- [X] [formatted your issue](https://yihui.org/issue/#please-format-your-issue-correctly) so it is easier for us to read?

- [difficult to do in this case, but provided what I could] included a minimal, self-contained, and reproducible example?

- [X] pasted the output from `xfun::session_info('bookdown')` in your issue?

- [I have the latest packages, and the latest major version of R and Rstudio, but not the latest minor version of R and Rstudio] upgraded all your packages to their latest versions (including your versions of R, the RStudio IDE, and relevant R packages)?

- [I'm using bookdown, not blogdown] installed and tested your bug with the development version of the blogdown package using `remotes::install_github()` ?

**Comment by cderv** [comment]
Thanks for opening a new issue. 

It seems you are using some HTML-only content in your book. This could happen with [htmlwidgets](https://bookdown.org/yihui/bookdown/html-widgets.html) or [web pages / shiny app](https://bookdown.org/yihui/bookdown/web-pages-and-shiny-apps.html). 

For those above , in order to build to PDF, a screenshot is taken and inserted in place. it seems that does not work for you. 

Which HTML specific content are you using in your book ? 

It is hard for us to help or look into an issue without a reproducible example.

**Comment by isaactpetersen** [comment]
Thanks very much for the prompt response. I use HTML content in several places (e.g., `paged_table()`, `kableExtra`, etc.), and it's hard for me to determine all of the instances of it across the book. Any suggestions for how I could identify all instances of HTML content?

In the meantime, I was able to generate a minimal reproducible example of the error:

If I add the following to `01-intro.rmd` of `bookdown-demo` (https://github.com/rstudio/bookdown-demo):

```{r}
paged_table(pressure)
```

I have tended to prefer `paged_table()` to `DT` because `paged_table()` allows paging across columns (in addition to rows) without a scroll bar. If `paged_table()` is the reason why my book isn't exporting to PDF, is there an easy way to use `paged_table()` for a given table if the output is HTML and to use `kable()` (or other) if the ouput is PDF? I'm fine to specify this on a table-by-table basis (rather than across all tables) because I don't use `paged_table()` for all tables.

There could also be other HTML content contributing to the error, too (e.g., `kableExtra` has had issues exporting to PDF: https://stackoverflow.com/questions/60011425/kableextra-error-in-markdown-when-rendering-to-pdf) . If I have to follow the suggestions to run `options(knitr.table.format = "latex")`, is it possible to specify that only if the output is PDF? If so, how could I do that?

Thanks again for your help!

**Comment by cderv** [comment]
Thank you for your precision. 

`rmarkdown::paged_table()` will produce a HTML table. It is not a htmlwidget like DT for example so it won't be specifically handle by knitr. knitr will do screenshot of the table in some specific case (for HTMLwidget table for example)

What you could do is have a `paged_table()` for your HTML output and another latex compatible table for your PDF output. You can make any chunk conditional to an output using some knitr function like `knitr::is_latex_output()`
See https://bookdown.org/yihui/rmarkdown-cookbook/latex-html.html

You can use that to customize your option depending on the case. About kableExtra, I was under the impression it should work with all format: https://haozhu233.github.io/kableExtra/bookdown/cross-format-tables-in-bookdown.html

We could also make **knitr** support paged table but this will end up with a screenshot of the table. Not sure you want that. 

Does the above solve your issue ?

**Comment by isaactpetersen** [comment]
That's very helpful and perfectly answers my question. That solves my issue, so feel free to close. Thanks very much for your help.

For others who have a similar question in the future, here's how I was able to get it to work in my case:

````
```{r eval = knitr::is_html_output(excludes = "epub")}
paged_table(mydata)
```

```{r eval = knitr::is_latex_output()}
kable(mydata)
```
````

**Comment by cderv** [comment]
Glad it works ! 

I think this is the solution to use when the table framework is HTML and you want a different output than screenshot in your pdf ! 

I'll close this issue.
